### PR TITLE
ROB: Cope with garbage collector during cloning

### LINKED ISSUE #1788: Random bug with add_page
With the same environment, i go to have a malformed pdf 

## Environment

```bash
$ python -m platform
Linux-5.19.0-38-generic-x86_64-with-Ubuntu-22.04-jammy

$ python -c "import pypdf;print(pypdf.__version__)"
3.7.1

```

## Code + PDF

```python            

import sys
from pypdf import PdfWriter, PdfReader                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                     
path_out = sys.argv[1]                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                    
output = PdfWriter()                                                                                                                                                                                                                                                                
l_id = []                                                                                                                                                                                                                                                                           
for idx in range(1, 26):                                                                                                                                                                                                                                                            
   r = PdfReader('test/%s.pdf' % idx)                                                                                                                                                                                                                                               
   l_id.append(str(id(r)))                                                                                                                                                                                                                                                          
   for p in r.pages:                                                                                                                                                                                                                                                                
       output.add_page(p)                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                    
output.write(path_out)                                                                                                                                                                                                                                                              
with open(path_out +'.txt', 'w') as fo:                                                                                                                                                                                                                                             
    fo.write("\n".join(l_id)) 
```
Create a directory test with different pdf 
Example pdf: https://file.io/Pn1IwMh9pc5y

I run the program, 50 times
`for i in $(seq 1 50); do python test_concat_pdf.py /tmp/result$i.pdf; done`

# Result
`md5sum /tmp/result*.pdf`
```bash
0ce2c9fb9f5fb8a16fee8c735a64e3a3  /tmp/result33.pdf
18ac040f2467156533b41cc45079436e  /tmp/result10.pdf
18ac040f2467156533b41cc45079436e  /tmp/result11.pdf
18ac040f2467156533b41cc45079436e  /tmp/result12.pdf
18ac040f2467156533b41cc45079436e  /tmp/result13.pdf
18ac040f2467156533b41cc45079436e  /tmp/result14.pdf
18ac040f2467156533b41cc45079436e  /tmp/result15.pdf
18ac040f2467156533b41cc45079436e  /tmp/result16.pdf
18ac040f2467156533b41cc45079436e  /tmp/result17.pdf
18ac040f2467156533b41cc45079436e  /tmp/result18.pdf
18ac040f2467156533b41cc45079436e  /tmp/result19.pdf
18ac040f2467156533b41cc45079436e  /tmp/result1.pdf
18ac040f2467156533b41cc45079436e  /tmp/result20.pdf
18ac040f2467156533b41cc45079436e  /tmp/result21.pdf
18ac040f2467156533b41cc45079436e  /tmp/result22.pdf
18ac040f2467156533b41cc45079436e  /tmp/result23.pdf
18ac040f2467156533b41cc45079436e  /tmp/result24.pdf
18ac040f2467156533b41cc45079436e  /tmp/result25.pdf
18ac040f2467156533b41cc45079436e  /tmp/result26.pdf
18ac040f2467156533b41cc45079436e  /tmp/result27.pdf
18ac040f2467156533b41cc45079436e  /tmp/result28.pdf
18ac040f2467156533b41cc45079436e  /tmp/result29.pdf
18ac040f2467156533b41cc45079436e  /tmp/result2.pdf
18ac040f2467156533b41cc45079436e  /tmp/result30.pdf
18ac040f2467156533b41cc45079436e  /tmp/result31.pdf
18ac040f2467156533b41cc45079436e  /tmp/result32.pdf
18ac040f2467156533b41cc45079436e  /tmp/result34.pdf
18ac040f2467156533b41cc45079436e  /tmp/result35.pdf
18ac040f2467156533b41cc45079436e  /tmp/result36.pdf
18ac040f2467156533b41cc45079436e  /tmp/result37.pdf
18ac040f2467156533b41cc45079436e  /tmp/result38.pdf
18ac040f2467156533b41cc45079436e  /tmp/result39.pdf
18ac040f2467156533b41cc45079436e  /tmp/result3.pdf
18ac040f2467156533b41cc45079436e  /tmp/result40.pdf
18ac040f2467156533b41cc45079436e  /tmp/result41.pdf
18ac040f2467156533b41cc45079436e  /tmp/result42.pdf
18ac040f2467156533b41cc45079436e  /tmp/result43.pdf
18ac040f2467156533b41cc45079436e  /tmp/result44.pdf
18ac040f2467156533b41cc45079436e  /tmp/result45.pdf
18ac040f2467156533b41cc45079436e  /tmp/result46.pdf
18ac040f2467156533b41cc45079436e  /tmp/result47.pdf
18ac040f2467156533b41cc45079436e  /tmp/result48.pdf
18ac040f2467156533b41cc45079436e  /tmp/result49.pdf
18ac040f2467156533b41cc45079436e  /tmp/result4.pdf
18ac040f2467156533b41cc45079436e  /tmp/result50.pdf
18ac040f2467156533b41cc45079436e  /tmp/result5.pdf
18ac040f2467156533b41cc45079436e  /tmp/result6.pdf
18ac040f2467156533b41cc45079436e  /tmp/result7.pdf
18ac040f2467156533b41cc45079436e  /tmp/result8.pdf
18ac040f2467156533b41cc45079436e  /tmp/result9.pdf
```
Here, a pdf is different. It has blank pages.
In the txt file, I have listed the ids of the PdfReader
For the wrong pdf, I have the same id twice (140495761167696)
```
140495777752976                                                                                                                                                                                                                                                                     
140495757290384
140495756671120
140495753673168
140495761167696
140495727106832
140495710644816
140495744436432
140495735690192
140495711730448
140495726845008
140495715144272
140495694350032
140495748724432
140495715062608
140495761167696
140495731812752
140495707371024
140495719364816
140495684779856
140495683795728
140495684528016
140495710266832
140495710765648
14049571560302
```

If I test an alternative script
```python
import sys                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                    
from pypdf import PdfWriter, PdfReader                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                    
path_out = sys.argv[1]                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                    
output = PdfWriter()                                                                                                                                                                                                                                                                
l_id = []                                                                                                                                                                                                                                                                           
l_r = []                                                                                                                                                                                                                                                                            
for idx in range(1, 26):                                                                                                                                                                                                                                                            
   r = PdfReader('test/%s.pdf' % idx)                                                                                                                                                                                                                                               
   l_r.append(r)    
   l_id.append(str(id(r)))    
   for p in r.pages:    
       output.add_page(p)    
    
output.write(path_out)    
with open(path_out +'.txt', 'w') as fo:    
    fo.write("\n".join(l_id))  
```
No problem. Adding the l_id list prevents duplicate ids

I would suggest adding a hash attribute on the PdfReader based on the stream attribute to replace the id method value. What do you think about it?

### PR DESCRIPTION
**Pull request by pubpub-zz**
fixes #1788

### CODE REVIEW

File: pypdf/generic/_base.py [hunk 1]
```diff
@@ -123,6 +123,7 @@ def _reference_clone(
         if ind is not None:
             if id(ind.pdf) not in pdf_dest._id_translated:
                 pdf_dest._id_translated[id(ind.pdf)] = {}
+                pdf_dest._id_translated[id(ind.pdf)][-10] = ind.pdf
```
Comment (MartinThoma) [inline pypdf/generic/_base.py:126]: Why -10? Doesn't this potentially override something important?

Comment (pubpub-zz) [reply pypdf/generic/_base.py:126]: No : negative ids are not used ; this mod only link the source pdf to the PdfWriter object, preventing its garbage collection and then therefore the id reuse. the GC will be performed when the source object will be reused

Comment (MartinThoma) [reply pypdf/generic/_base.py:126]: What does the `-10`mean? I've just noticed that this is a dictionary, not a list. Does -10 have  any special meaning?

Comment (pubpub-zz) [reply pypdf/generic/_base.py:126]: Nothing special I selected a number far enough to leave immediate negative numbers available for other use

**Comment by pubpub-zz** [comment]
+1

**Comment by pubpub-zz** [comment]
oups forgot to say all Yours!🤣

**Comment by MartinThoma** [comment]
I'm still uncertain about the `-10`. It's just so random. I don't understand where that comes from / what effect this change has.

**Comment by pubpub-zz** [comment]
if you prefer another magic number, I have no preference.

**Comment by pubpub-zz** [comment]
you may prefer this proposal🙂🙂

**Comment by pubpub-zz** [comment]
+1?

**Comment by MartinThoma** [comment]
@pubpub-zz I've decided to simply trust you on this one :smile: I don't understand what this does / how this is fixing the issue, but you have proven to know a lot about pypdf + PDF specs. Hence I feel comfortable with merging it :hugs:

**Pull request merged**